host/cli/volume.go

```diff
@@ -0,0 +1,210 @@
+       flynn-host volume delete ID...
+       flynn-host volume gc
+
+Commands:
+    list    Display a list of all volumes of known Flynn hosts
+    create  Creates a volume on a host
+    delete  Deletes volumes, destroying any data stored on them
+    gc      Garbage collect currently unused volumes
+
+Examples:
+
+    $ flynn-host volume list
+
+    $ flynn-host volume create --provider zfs host0
+
+    $ flynn-host volume destroy 102fad07-07a3-4841-bded-d9e8a3eedbd6
+
+    $ flynn-host volume gc
+`)
+}
+
+func runVolume(args *docopt.Args, client *cluster.Client) error {
+	switch {
+	case args.Bool["list"]:
+		return runVolumeList(args, client)
+	case args.Bool["delete"]:
+		return runVolumeDelete(args, client)
+	case args.Bool["create"]:
+		return runVolumeCreate(args, client)
+	case args.Bool["gc"]:
+		return runVolumeGarbageCollection(args, client)
+	}
+	return nil
+}
+
+func runVolumeGarbageCollection(args *docopt.Args, client *cluster.Client) error {
+	// collect list of all volume ids currently attached to jobs
+	hosts, err := client.Hosts()
+	if err != nil {
+		return fmt.Errorf("could not list hosts: %s", err)
+	}
+	if len(hosts) == 0 {
+		return errors.New("no hosts found")
+	}
+
+	attached := make([]string, 0, 0)
+	for _, h := range hosts {
+		jobs, err := h.ListJobs()
+		if err != nil {
+			return err
+		}
+		for _, j := range jobs {
+			for _, vb := range j.Job.Config.Volumes {
+				attached = append(attached, vb.VolumeID)
+			}
+		}
+	}
+
+	volumes, err := clusterVolumes(hosts)
+	if err != nil {
+		return err
+	}
+
+	// iterate over list of all volumes, deleting any not found in the attached list
+	success := true
+outer:
+	for _, v := range volumes {
+		for _, a := range attached {
+			if a == v.Volume.ID {
+				// volume is attached, continue to next volume
+				continue outer
+			}
+		}
+		if err := v.Host.DestroyVolume(v.Volume.ID); err != nil {
+			success = false
+			fmt.Printf("could not delete volume %s: %s\n", v.Volume.ID, err)
+			continue outer
+		}
+		fmt.Println(v.Volume.ID, "deleted")
+	}
+	if !success {
+		return errors.New("could not garbage collect all volumes")
+	}
+
+	return nil
+}
+
+func runVolumeDelete(args *docopt.Args, client *cluster.Client) error {
+	success := true
+	hosts, err := client.Hosts()
+	if err != nil {
+		return fmt.Errorf("could not list hosts: %s", err)
+	}
+	if len(hosts) == 0 {
+		return errors.New("no hosts found")
+	}
+
+	volumes, err := clusterVolumes(hosts)
+	if err != nil {
+		return err
+	}
+
+outer:
+	for _, id := range args.All["ID"].([]string) {
+		// find this volume in the list
+		for _, v := range volumes {
+			if v.Volume.ID == id {
+				if err := v.Host.DestroyVolume(id); err != nil {
+					success = false
+					fmt.Printf("could not delete volume %s: %s\n", id, err)
+					continue outer
+				}
+				// delete the volume
+				fmt.Println(id, "deleted")
+				continue outer
+			}
+		}
+		success = false
+		fmt.Printf("could not delete volume %s: volume not found\n", id)
```

Comment:
Unfortunately because Volume IDs don't encode the host we can't lookup the host directly, so if we can't find the volume in the list of volumes we were able to enumerate we can't attempt to delete it, as we don't know where it was/is located.